File: one_million/optimized/one_million.go
package main

import (
	"strconv"
	"syscall"
)

func main() {
	const bufSize = 65536
	var buf [bufSize]byte
	pos := 0
	prefix := []byte("Hello, this is iteration number: ")
	prefixLen := len(prefix)
	var numBuf [16]byte

	for i := 0; i < 1000000; i++ {
		if pos+prefixLen+16 > bufSize {
			_, _ = syscall.Write(1, buf[:pos])
			pos = 0
		}
		copy(buf[pos:], prefix)
		pos += prefixLen
		formatted := strconv.AppendInt(numBuf[:0], int64(i), 10)
		copy(buf[pos:], formatted)
		pos += len(formatted)
		buf[pos] = '\n'
		pos++
	}
	if pos > 0 {
		_, _ = syscall.Write(1, buf[:pos])
	}
}

File: one_million/optimized/one_million.c
#include <unistd.h>
#include <string.h>

#define BUF_SIZE (64 * 1024)

int main(void) {
	char buffer[BUF_SIZE];
	size_t pos = 0;
	const char prefix[] = "Hello, this is iteration number: ";
	const size_t prefix_len = sizeof(prefix) - 1;

	for (int n = 0; n < 1000000; n++) {
		if (pos + prefix_len + 12 > BUF_SIZE) {
			if (write(1, buffer, pos) < 0) return 1;
			pos = 0;
		}
		memcpy(buffer + pos, prefix, prefix_len);
		pos += prefix_len;

		char num_buf[10];
		int temp = n;
		int digits = 0;
		if (temp == 0) {
			num_buf[digits++] = '0';
		} else {
			while (temp > 0) {
				num_buf[digits++] = (char)('0' + (temp % 10));
				temp /= 10;
			}
		}
		for (int j = digits - 1; j >= 0; j--) {
			buffer[pos++] = num_buf[j];
		}
		buffer[pos++] = '\n';
	}
	if (pos > 0) {
		if (write(1, buffer, pos) < 0) return 1;
	}
	return 0;
}
